libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkZstdLibrary.java

```diff
@@ -0,0 +1,99 @@
+import static java.lang.foreign.ValueLayout.JAVA_BOOLEAN;
+import static java.lang.foreign.ValueLayout.JAVA_BYTE;
+import static java.lang.foreign.ValueLayout.JAVA_INT;
+import static java.lang.foreign.ValueLayout.JAVA_LONG;
+import static org.elasticsearch.nativeaccess.jdk.LinkerHelper.downcallHandle;
+
+public class JdkZstdLibrary implements ZstdLibrary {
+
+    static {
+        System.loadLibrary("zstd");
+    }
+
+    private static final MethodHandle compressBound$mh = downcallHandle("ZSTD_compressBound", FunctionDescriptor.of(JAVA_LONG, JAVA_INT));
+    private static final MethodHandle compress$mh = downcallHandle("ZSTD_compress", FunctionDescriptor.of(JAVA_LONG, ADDRESS, JAVA_INT, ADDRESS, JAVA_INT, JAVA_INT));
+    private static final MethodHandle isError$mh = downcallHandle("ZSTD_isError", FunctionDescriptor.of(JAVA_BOOLEAN, JAVA_LONG));
+    private static final MethodHandle getErrorName$mh = downcallHandle("ZSTD_getErrorName", FunctionDescriptor.of(ADDRESS, JAVA_LONG));
+    private static final MethodHandle decompress$mh = downcallHandle("ZSTD_decompress", FunctionDescriptor.of(JAVA_LONG, ADDRESS, JAVA_INT, ADDRESS, JAVA_INT));
+
+    @Override
+    public long compressBound(int srcLen) {
+        try {
+            return (long) compressBound$mh.invokeExact(srcLen);
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
+    }
+
+    @Override
+    public long compress(ByteBuffer dst, int dstLen, ByteBuffer src, int srcLen, int compressionLevel) {
+        try (Arena arena = Arena.ofConfined()) {
+            var nativeDst = arena.allocate(dstLen);
+            var nativeSrc = arena.allocate(srcLen);
+            nativeSrc.asByteBuffer().put(0, src, src.position(), srcLen);
+            var compressedLen = (long) compress$mh.invokeExact(nativeDst, dstLen, nativeSrc, srcLen, compressionLevel);
+            if (isError(compressedLen) == false) {
+                assert compressedLen <= dstLen;
+                dst.put(dst.position(), nativeDst.asByteBuffer(), 0, (int) compressedLen);
+            }
+            return compressedLen;
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
+    }
+
+    @Override
+    public boolean isError(long code) {
+        try {
+            return (boolean) isError$mh.invokeExact(code);
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
+    }
+
+    @Override
+    public String getErrorName(long code) {
+        try {
+            MemorySegment str = (MemorySegment) getErrorName$mh.invokeExact(code);
+            return str.reinterpret(Long.MAX_VALUE).getUtf8String(0);
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
+    }
+
+    @Override
+    public long decompress(ByteBuffer dst, int dstLen, ByteBuffer src, int srcLen) {
+        try (Arena arena = Arena.ofConfined()) {
+            var nativeDst = arena.allocate(dstLen);
```

Comment:
Likewise here for decompression, maybe we should take advantage of the `ZSTD_getFrameContentSize` API to avoid allocating more native memory than necessary.

Reply:
The `compress` method that is exposed right now is guaranteed to only create a single frame: https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L150 so that should always be the case, but you have a point that we may want to decompress buffers compressed by 3rd parties in the future.

We could check if the source length is equal to `ZSTD_findFrameCompressedSize` to check if there is a single frame. But at this point this is becoming complicated enough that it's probably worth deferring to a follow-up PR.